File: datastructure/union-find/atcoder/atc_B.py
class UnionFind:

    def __init__(self, n):
        # -1で初期化する
        self.parents = [-1] * n
        # 木の高さを格納する
        self.rank = [0] * n
    
    # xの根を探す
    def find(self, x):
        if self.parents[x] == -1:
            return x
        else:
            # 経路圧縮
            self.parents[x] = self.find(self.parents[x])
            return self.parents[x]
    
    # xとyの属する集合を併合する
    def unite(self, x, y):
        # x, yの根を探す
        x = self.find(x)
        y = self.find(y)

        # 同じ集合に属している場合
        if x == y:
            return
        
        # 高さが低い方を高い方に併合する
        if self.rank[x] < self.rank[y]:
            self.parents[x] = y
        else:
            self.parents[y] = x
            # 高さが同じ場合は高さを1増やす
            if self.rank[x] == self.rank[y]:
                self.rank[x] += 1
    
    # xとyが同じ集合に属するか判定する
    def same(self, x, y):
        return self.find(x) == self.find(y)


if __name__ == '__main__':
    n, q = map(int, input().split())
    uf = UnionFind(n)
    for _ in range(q):
        p, a, b = map(int, input().split())
        if p == 0:
            uf.unite(a, b)
        else:
            if uf.same(a, b):
                print('Yes')
            else:
                print('No')

File: datastructure/union-find/atcoder/atc_B.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strconv"
)

type UnionFind struct {
	parents []int
	ranks   []int
}

func NewUnionFind(n int) *UnionFind {
	parents := make([]int, n)
	ranks := make([]int, n)
	for i := 0; i < n; i++ {
		parents[i] = i
		ranks[i] = 1
	}
	return &UnionFind{parents, ranks}
}

func (uf *UnionFind) Find(x int) int {
	if uf.parents[x] != x {
		uf.parents[x] = uf.Find(uf.parents[x])
	}
	return uf.parents[x]
}

func (uf *UnionFind) Union(x, y int) {
	xRoot, yRoot := uf.Find(x), uf.Find(y)
	if xRoot == yRoot {
		return
	}
	if uf.ranks[xRoot] < uf.ranks[yRoot] {
		uf.parents[xRoot] = yRoot
	} else if uf.ranks[xRoot] > uf.ranks[yRoot] {
		uf.parents[yRoot] = xRoot
	} else {
		uf.parents[yRoot] = xRoot
		uf.ranks[xRoot]++
	}
}

func (uf *UnionFind) IsConnected(x, y int) bool {
	return uf.Find(x) == uf.Find(y)
}

func main() {
    scanner := bufio.NewScanner(os.Stdin)
    scanner.Split(bufio.ScanWords)

    scanner.Scan()
    n, _ := strconv.Atoi(scanner.Text())

    scanner.Scan()
    q, _ := strconv.Atoi(scanner.Text())

    uf := NewUnionFind(n)

    for i := 0; i < q; i++ {
        scanner.Scan()
        p, _ := strconv.Atoi(scanner.Text())

        scanner.Scan()
        a, _ := strconv.Atoi(scanner.Text())

        scanner.Scan()
        b, _ := strconv.Atoi(scanner.Text())

        if p == 0 {
            uf.Union(a, b)
        } else {
            if uf.IsConnected(a, b) {
                fmt.Println("Yes")
            } else {
                fmt.Println("No")
            }
        }
    }
}
